packages/host/app/services/loader-service.ts

```diff
@@ -47,68 +51,34 @@ export default class LoaderService extends Service {
       new URL(baseRealm.url),
       new URL(config.resolvedBaseRealmURL),
     );
-    loader.prependURLHandlers([(req) => this.fetchWithAuth(req)]);
+    loader.prependURLHandlers([(req) => this.addAuthorizationHeader(req)]);
     shimExternals(this.virtualNetwork);
 
     return loader;
   }
 
-  private async fetchWithAuth(request: Request) {
-    // To avoid deadlock, we can assume that any GET requests to baseRealm don't need authentication.
-    let isGetRequestToBaseRealm =
-      request.url.includes(baseRealm.url) && request.method === 'GET';
-    if (
-      isGetRequestToBaseRealm ||
-      request.url.endsWith('_session') ||
-      request.method === 'HEAD' ||
-      request.headers.has('Authorization')
-    ) {
-      return null;
-    }
-    let realmURL = await this.realmInfoService.fetchRealmURL(request.url);
-    if (!realmURL) {
-      return null;
-    }
-
-    // We have to get public readable status
-    // before we instantiate realm resource and load realm token.
-    // Because we don't want to do authentication
-    // for GET request to publicly readable realm.
-    let isPublicReadable = await this.realmInfoService.isPublicReadable(
-      realmURL,
-    );
-    if (request.method === 'GET' && isPublicReadable) {
-      return null;
-    }
-
-    await this.matrixService.ready;
-    if (!this.matrixService.isLoggedIn) {
-      return null;
-    }
-
-    let realmSession = await this.getRealmSession(realmURL);
-    if (!realmSession.rawRealmToken) {
-      return null;
-    }
-
-    request.headers.set('Authorization', realmSession.rawRealmToken);
-
-    let response = await this.loader.fetch(request);
-
-    // We will get a 401 from the server in three cases. When the token is invalid,
-    // the JWT is expired, and the permissions in the JWT payload differ
-    // from what is configured on the server (e.g. someone changed permissions for the user during the life
-    // of the JWT). In those cases, we have to retrieve a new JWT.
-    if (!response.ok && response.status === 401) {
-      await realmSession.refreshToken();
-      if (!realmSession.rawRealmToken) {
-        return null;
-      }
-      request.headers.set('Authorization', realmSession.rawRealmToken);
-      response = await this.loader.fetch(request);
-    }
-
-    return response;
+  private async addAuthorizationHeader(request: Request) {
+    return await addAuthorizationHeader(this.loader, request, {
+      getJWT: async (realmURL: string) => {
+        return (await this.getRealmSession(new URL(realmURL))).rawRealmToken!;
+      },
+      getRealmInfoByURL: async (url: string) => {
```

Comment:
We can drop `byURL` postfix, it's already evident from the argument. `getRealmInfo(url)` is enough